enhancements/authentication/pod-security-admission-enforcement.md

```diff
@@ -0,0 +1,315 @@
+- If the `FeatureGate` `OpenShiftPodSecurityAdmission` is removed from the current `FeatureSet`, the cluster must revert to its previous state.
+  It serves as a break-glass mechanism.
+
+#### PodSecurityReadinessController
+
+The `PodSecurityReadinessController` will manage the `PSAEnforcementConfig` API.
+It already collects most of the necessary data to determine whether a Namespace would fail enforcement or not to create a [`ClusterFleetEvaluation`](https://github.com/openshift/enhancements/blob/master/dev-guide/cluster-fleet-evaluation.md).
+With the `security.openshift.io/MinimallySufficientPodSecurityStandard`, it will be able to evaluate all Namespaces for failing workloads, if any enforcement would happen.
+With the `security.openshift.io/ValidatedSCCSubjectType`, it can categorize violations more accurately and create a more accurate `ClusterFleetEvaluation`.
+
+#### PodSecurity Configuration
+
+A Config Observer in the `cluster-kube-apiserver-operator` manages the Global Config for the kube-apiserver, adjusting behavior based on the `OpenShiftPodSecurityAdmission` `FeatureGate`.
+It must watch both the `status.enforcementMode` for `Restricted` and the `FeatureGate` `OpenShiftPodSecurityAdmission` to be enabled to make decisions.
+
+#### PodSecurityAdmissionLabelSynchronizationController
+
+The [PodSecurityAdmissionLabelSynchronizationController (PSA label syncer)](https://github.com/openshift/cluster-policy-controller/blob/master/pkg/psalabelsyncer/podsecurity_label_sync_controller.go) must watch the `status.enforcementMode` and the `OpenShiftPodSecurityAdmission` `FeatureGate`.
+If `spec.enforcementMode` is `Restricted` and the `FeatureGate` `OpenShiftPodSecurityAdmission` is enabled, the syncer will set the `pod-security.kubernetes.io/enforce` label.
+Otherwise, it will refrain from setting that label and remove any enforce labels it owns if existent.
+
+Because the ability to set `pod-security.kubernetes.io/enforce` is introduced, the ability to remove that label must exist in the release before.
+Otherwise, the cluster will be unable to revert to its previous state.
+
+## Open Questions
+
+### Fresh Installs
+
+Needs to be evaluated. The System Administrator needs to pre-configure the new API’s `spec.enforcementMode`, choosing whether the cluster will be `Privileged` or `Restricted` during a fresh install.
+
+### Impact on HyperShift
+
+Needs to be evaluated.
+
+### Baseline Clusters
+
+The current suggestion differentiates between `Restricted` and `Privileged` PSS.
+It may be possible to introduce an intermediate step and set the cluster to `baseline` instead.
+
+### Enforce PSA labe syncer, fine-grained
+
+It would be possible to enforce only the `pod-security.kubernetes.io/enforce` labels on Namespaces without enforcing it globally through the `PodSecurity` configuration given to the kube-apiserver.
```

Comment:
How would this look in reality? Are there race issues with a solution that leverages this?

Reply:
Oh, yes. We need to actually do this. We need a signal from the PSA label syncer, that it is done, before we start enforcing the global configuration. The question is, if it is better to spread out to another release or if we would like to introduce a signal of state from the PSA label syncer.

Why we need to coordinate this:
- Workloads would fail (if they need baseline / privileged) for the duration between global configuration enforcing restricted and the label syncer setting an enforce label that is less restrictive than `restricted`.

Reply:
Does the enforce label not have a value that matches the privileged value today? If the label syncer were to sync a value that demonstrates the default behaviour ahead of the pivot, the it strikes me the race could be mitigated and only later would the label get updated?

Or does the label syncer abandon namespaces once labelled and so by labelling it privileged, it would no longer be able to later label it restricted?

Reply:
That is a great idea. I will try to discuss this in today's Arch Call.

The abandon logic is something that I need to verify isn't the case as it will need to revisit all Namespaces it touched once it will need to enforce.